docs/proposals/20210222-machineadm.md

```diff
@@ -0,0 +1,835 @@
+</td>
+</tr>
+
+<tr>
+<td>machineadm remote configuration plugin</td><td>infrastructure provider repo</td>
+<td>
+An infrastructure plugin for the node bootstrapper will be created within infrastructure providers that would be responsible for infrastructure-specific logic to pull files machineadm files from a given location and upload
+bootstrap data.
+</td>
+</tr>
+
+<tr>
+<td>machineadm encryption plugin</td><td>infrastructure provider repo</td>
+<td>
+An infrastructure plugin for the node bootstrapper will be created within infrastructure providers that would be responsible for infrastructure-specific logic to use an external key management service.
+</td>
+</tr>
+
+<tr>
+<td>machineadm infrastructure controller</td><td>infrastructure provider repo</td>
+<td>
+A controller that ships with the infrastructure provider that can reconcile the storage of machineadm configurations with infrastructure APIs (e.g. Amazon S3/Minio, GCS, custom server etc...).
+</td>
+</tr>
+
+</tbody>
+</table>
+
+##### Core plugins
+<table>
+<thead>
+<tr><th>Component</th><th>Description</th>
+</thead>
+<tbody>
+
+<tr>
+<td>machineadm-container-runtime-plugin-containerd</td>
+<td>
+A plugin that configures containerd runtimes.
+</td>
+
+<tr>
+<td>machineadm-timesync-plugin-systemd</td>
+<td>
+A plugin that configures systemd-timesyncd.
+</td>
+
+<tr>
+<td>machineadm-timesync-plugin-ntpd</td>
+<td>
+A plugin that configures ntpd.
+</td>
+
+<tr>
+<td>machineadm-timesync-plugin-chronyd</td>
+<td>
+A plugin that configures chronyd.
+</td>
+
+</tbody>
+</table>
+
+### Phased development
+
+With reference to the modalities described in the requirements specification, we propose 3 phases of development:
+
+| Phase             | What gets implemented           |
+| ----------------- | ------------------------------- |
+| Phase 1 - Alpha 1 | Provisioning Modality           |
+| Phase 2 - Alpha 2 | Preparation and Post Modalities |
+| Phase 3 & GA      | API Stabilisation               |
+
+### Plugin architecture
+
+Machineadm will use the [go-plugin][go-plugin] architecture used by Hashicorp. Machineadm will expect plugins to
+use consistent filename formats depending on what type of plugin it is (and will be documented in the type).
+
+Plugins can either be "on path", in a directory supported in the PATH variable, or:
+* Linux: stored in `/usr/libexec/machineadm` or `/usr/local/libexec/machineadm`
+* Windows: stored in `"C:\Program Files\machineadm\plugins"`
+
+### Security Model
+
+#### Userdata storage
+
+Machineadm's security model is intended such that secrets associated with machine configuration are stored on the
+management Kubernetes cluster and are secured to the level that the backing API server and etcd are secured.
+
+For machine bootstrapping, secrets are intended to be delivered by an infrastructure provider to a suitable secure
+location, and to also provide a mechanism for the machine to report back status. Examples include:
```

Comment:
This seems to be a very weak model considering the pre-requisites for setting up encryption, and the lack of audibility (Due to the verbosity of Kubernetes auditing and lack of etcd auditing)

Why not just rely on the cloud providers for this through something like sops.

machineadm.yml -> sops encypt -> MachineConfig -> S3
S3 -> sops decrypt | machineadm -c -

sops is widely tested, easily usable as library and supports AWS, GCP, Azure, Hashicorp and PGP

Reply:
Noted. The envelope format is one I've used before based off the old Python variant of SOPS. Noted, SOPS is now in Golang, so it's worth a revisit. Thanks.

Reply:
From a decryption perspective using it as a library just works: https://pkg.go.dev/go.mozilla.org/sops/v3@v3.6.1/decrypt#Data

It will then use the SOPS metadata in the metadata to try all available decryption mechanisms. 

For encryption, it is interfaced using MasterKey, which are creatable with a single function call to the KMS provider e.g.  https://pkg.go.dev/go.mozilla.org/sops/v3@v3.6.1/kms#NewMasterKeyFromArn

Reply:
SGTM. I'll take a look and will update.